### ISSUE TITLE
About visualization of the attention map

### ISSUE THREAD

**Issue by JialeTao** [leeyegy/TokenPose#3]
Hi, thanks for sharing the good work. I have some questions about the Fig. 3 in the paper.

![image](https://user-images.githubusercontent.com/38600167/137494151-337b6b0c-5484-418c-86b1-6cf73905726f.png)

1. Note that you use the TokenPose-S model to visualize the attention maps, are the situations similar with that of the TokenPoseB and TokenPoseL model? I post my understanding and I don't know if it is right: It seems when use the TokenPoseS model, the CNN is small and the feature can be related or matched to the original image. While turned to a large CNN model such as HRNet-W32, the feature map may be more high-level, did you visualize the attention map of this situation?
 
2. The output feature map of the CNN model is downsampled by 4, and the patch feature is further dowsampled with patch size 4*4 when computing attention. So the attention map is 16 times smaller than the original image, how do you transform the attention map to the image size? Direct resize or some operations like repeat the patch attention map?

3. Which result and method you used for visualization? I mean if you use the attention map after softmax operation for visualization? Or before softmax operation, use some orther normlization methods for normalizing the dot results? I think the softmax will cause the single peak problem, while in Fig. 3, the obvious region attentions can be observerd.

Any replys will be appreciated! And it will be more helpful if you can share the visualization code!

**Comment by yangsenius** [comment]
Hi, @JialeTao 

1. We did not visualize the attention maps of the TokenPose-B model. But we did visualize the TokenPose-Large model with keypoint token fusion. And the situation is different from the TokenPose-S model and we describe it in the paper. The keypoint fusion of TokenPose-L will enable the final keypoint representation to directly exploit the attention information in the early layers such that the attention patterns in the latter layers show less clear evidence for the keypoint localization. For the HRNet-based TokenPose without keypoint fusion, we suggest that you can refer to the visualization analysis of [TransPose](https://github.com/yangsenius/TransPose) on the HRNet based model. Multi-scale fusion in HRNet does cause e a different attention pattern.

2. We use bilinear interpolation to upsample the original attention map size to the image size.

3. We use the hook function of pytorch to hook the queries and keys, compute the normalized dot-product attention, and visualize the attention maps after the softmax operation. Yes, the softmax will make the distribution more extreme.

**Comment by JialeTao** [comment]
Hi, thanks for detail reply. I have little confusion now, only about the situation in the TokenPose-B and -L model. I have some guess but no envidence, and thanks for suggesting the paper, I'm going to read.
